src/main/java/org/opentripplanner/middleware/tripmonitor/jobs/CheckMonitoredTrip.java

```diff
@@ -498,6 +549,95 @@ public TripMonitorNotification checkTripForDelay(NotificationType delayType) {
+        NotificationType delayType,
+        List<Itinerary> itineraries,
+        int nextLegIndex
+    ) {
+        var itineraryToCheck = getMatchingItinerary(delayType, itineraries);
+        if (itineraryToCheck != null) {
+            int majorChangeIndex = IntStream
+                .range(0, itineraryToCheck.legs.size())
+                .filter(i -> hasMajorChange(i, delayType, itineraryToCheck))
+                .findFirst()
+                .orElse(-1);
+
+            if (majorChangeIndex == -1 || majorChangeIndex <= (nextLegIndex - 1)) {
+                // No major change or the traveler has already passed it.
+                return null;
+            }
+
+            // Found a major change.
+            switch (delayType) {
+                case MODE_CHANGE:
+                    return TripMonitorNotification.createModeChangeNotification(
+                        trip.itinerary.legs.get(majorChangeIndex),
+                        itineraryToCheck.legs.get(majorChangeIndex),
+                        getOtpUserLocale()
+                    );
+                case ORIGIN_CHANGE:
+                case DESTINATION_CHANGE:
+                    return TripMonitorNotification.createStopChangeNotification(
+                        delayType,
+                        trip.itinerary.legs.get(majorChangeIndex),
+                        itineraryToCheck.legs.get(majorChangeIndex),
+                        getOtpUserLocale()
+                    );
+                default:
+                    return null;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Check the appropriate leg for a major change based on the provided exception.
+     */
+    private boolean hasMajorChange(int index, NotificationType exception, Itinerary itineraryToCheck) {
+        if (exception == NotificationType.MODE_CHANGE) {
+            return !ItineraryUtils.legsModeMatches(trip.itinerary.legs.get(index), itineraryToCheck.legs.get(index));
+        } else if (exception == NotificationType.ORIGIN_CHANGE){
+            return !ItineraryUtils.stopsMatch(trip.itinerary.legs.get(index).from, itineraryToCheck.legs.get(index).from);
+        } else if (exception == NotificationType.DESTINATION_CHANGE){
+            return !ItineraryUtils.stopsMatch(trip.itinerary.legs.get(index).to, itineraryToCheck.legs.get(index).to);
+        }
+        return false;
+    }
+
+    /**
+     * A list of all notification types that are classed as major changes to an itinerary.
+     */
+    private List<NotificationType> getAllMajorChangeNotificationTypes() {
+        return List.of(NotificationType.MODE_CHANGE, NotificationType.ORIGIN_CHANGE, NotificationType.DESTINATION_CHANGE);
+    }
+
+    /**
+     * Attempt to match on just a single exception, else use the wider exception criteria.
+     */
+    @Nullable
+    private Itinerary getMatchingItinerary(NotificationType delayType, List<Itinerary> itineraries) {
```

Comment:
Rename `delayType` to `notificationType`

Reply:
Refactored and addressed in: https://github.com/ibi-group/otp-middleware/pull/275/commits/91d66190cb611283993b8e745148641efd695c5e